Sample mean transcript coverage over a list of genes

northwestwitch opened the pull request:
### This PR adds | fixes:
- Return coverage over a list of genes (transcripts only) for a sample in the database

### How to test:
- Launch an instance of the app and provide a sample with a d4 file and a list of genes in various formats (ensembl ids, hgnc ids, hgnc symbols)

### Expected outcome:
- [x] Mean transcript coverage should be returned for each provided gene

### Review:
- [ ] Code approved by
- [x] Tests executed by CR

This [version](https://semver.org/) is a:
- [ ] **MAJOR** - when you make incompatible API changes
- [x] **MINOR** - when you add functionality in a backwards compatible manner
- [ ] **PATCH** - when you make backwards compatible bug fixes or documentation/instructions

northwestwitch commented:
Local tests with a sample with a real d4 file:

- By providing a list of HGNC symbols: 🆗 

<img width="537" alt="image" src="https://user-images.githubusercontent.com/28093618/235908177-4cf97310-04b7-443e-92b2-1aab3bc4a05a.png">

- By providing a list of HGNC IDs: 🆗 

<img width="475" alt="image" src="https://user-images.githubusercontent.com/28093618/235908501-d22ac0ea-0ebb-4d69-92e1-98bff3ba0f57.png">

- By providing a list of Ensembl IDs: 🆗 

<img width="479" alt="image" src="https://user-images.githubusercontent.com/28093618/235908843-072a7851-7152-40c5-8a90-acbb5013b591.png">

northwestwitch commented:
This one is ready for review

henrikstranneheim commented on src/chanjo2/meta/handle_d4.py:
```diff
@@ -58,3 +62,42 @@ def genes_coverage(d4_file: D4File, genes: List[SQLGene]) -> List[CoverageInterv
+            limit=None,
+        )
+        mean_gene_cov: float = 0
+        if gene_transcripts:
+            mean_gene_cov: float = mean(
+                d4_file.mean(
+                    [
+                        (transcript.chromosome, transcript.start, transcript.stop)
+                        for transcript in gene_transcripts
+                    ]
+                )
+            )
```
I think this part would be good as a function:
mean_gene_coverage: float = get_mean_gene_coverage(args)
Maybe we can make it general enough to reuse for exons.

northwestwitch replied:
Done

henrikstranneheim commented on src/chanjo2/models/pydantic_models.py:
```diff
@@ -133,10 +133,10 @@ class CoverageInterval(BaseModel):
     ensembl_gene_id: Optional[str]
     hgnc_id: Optional[int]
     hgnc_symbol: Optional[str]
+    interval_id: Optional[int]
```
Would be nice to order this alphabetically to make it easier to scan

henrikstranneheim commented on tests/src/chanjo2/endpoints/test_coverage.py:
```diff
@@ -163,23 +163,23 @@ def test_d4_intervals_coverage(
     """Test the function that returns the coverage over multiple genes of a sample when a list of HGNC symbols is provided."""
 
-    # GIVING a sample coverage query containing HGNC gene symbols
+    # GIVING a sample gene coverage query containing HGNC gene symbols
```
```suggestion
    # GIVEN a sample gene coverage query containing HGNC gene symbols
```
Fix for all occurrences

henrikstranneheim commented on tests/src/chanjo2/endpoints/test_coverage.py:
```diff
@@ -215,23 +215,107 @@ def test_samples_gene_coverage_hgnc_ids(
+    assert response.status_code == status.HTTP_200_OK
+
+    # AND return coverage intervals data
+    result = response.json()
```
```suggestion
    coverage_intervals: list = response.json()
```

northwestwitch replied:
Done

henrikstranneheim commented on tests/src/chanjo2/endpoints/test_coverage.py:
```diff
@@ -215,23 +215,107 @@ def test_samples_gene_coverage_hgnc_ids(
+    assert response.status_code == status.HTTP_200_OK
+
+    # AND return coverage intervals data
+    result = response.json()
```
Same here.

henrikstranneheim commented on tests/src/chanjo2/endpoints/test_coverage.py:
```diff
@@ -215,23 +215,107 @@ def test_samples_gene_coverage_hgnc_ids(
+    endpoints: Type,
+    genomic_ids_per_build: Dict[str, List],
+):
+    """Test the function that returns the coverage over  transcripts of multiple genes of a sample when a list of Enseml IDs is provided."""
```
```suggestion
    """Test the function that returns the coverage over transcripts of multiple genes of a sample when a list of Enseml IDs is provided."""
```

henrikstranneheim reviewed (approved):
Fantastic 👍. It is getting more and more complete

northwestwitch commented on src/chanjo2/meta/handle_d4.py:
```diff
@@ -58,3 +62,42 @@ def genes_coverage(d4_file: D4File, genes: List[SQLGene]) -> List[CoverageInterv
     return genes_cov
+
+
+def genes_transcript_coverage(db: Session, d4_file: D4File, genes: List[SQLGene]):
```
Done

northwestwitch commented:
Ok, I think I have addressed all your comments @henrikstranneheim. I would like to integrate the completeness also in the transscripts stats, but I'd rather do it in another PR, so it's less messy!
